### DISCUSSION TITLE
Why did PySPH skip some syntax when executing Equation？

### DISCUSSION THREAD

**Discussion by hrbyrb** [pypr/pysph#349] [Q&A] [answered]
I was trying to write a new set of constitutive equations, but when I executed the example, I found a line of code in the **FintDragForce4(equation)** that never executed. As shown in Figure1:

![418a4aa69181dfcdd5a3aafe0d617ca](https://user-images.githubusercontent.com/73709907/170258011-29f17983-8fc3-4eb3-847b-81ab681100a2.png)
**Figure 1**

I checked the.hdF5 data file, in which the betaIMEC value is 0, but none of the parameters in the formula is 0, and I have executed this calculation code externally, and got its specific value 839.7, as shown in Figure 2 and 3.
![betaIMEC](https://user-images.githubusercontent.com/73709907/170259988-57bc4fe0-a852-498c-8201-1789f4a531d9.jpg)
**Figure 2**

![betaIMEC_test](https://user-images.githubusercontent.com/73709907/170260790-8b141445-01fb-4af2-bac4-328de01058fb.jpg)
**Figure 3**

Also I have added all the variables to the util file, so a betaIMEC value of 0 should not be without output.
![image](https://user-images.githubusercontent.com/73709907/170261828-9bba4317-4860-4096-b983-480f74510f5c.png)

- **Has anyone had any problems like this, or can anyone offer any suggestion?**

The **FintDragForce4(equation)** code is as follows：

```
class FintDragForce4(Equation):
    r"""Calculate the drag force, including the interphase momentum exchange 
        coefficient(beta_IMEC)

    """
    def __init__(self, dest, sources, rho=1000.0, nu=0.02, 
                 gx=0.0, gy=0.0, gz=0.0, dim=2):
        self.rho = rho
        self.nu = nu
        self.dim = dim
        self.gx = gx
        self.gy = gy
        self.gz = gz
        super(FintDragForce4, self).__init__(dest, sources)

    def initialize(self, d_idx, d_Fdrau, d_Fdrav, d_Fdraw, 
                   d_Fintu, d_Fintv, d_Fintw, d_betaIMEC, d_Re, d_Cdrag):
        
        d_betaIMEC[d_idx] = 0.0
        d_Re[d_idx] = 0.0
        d_Cdrag[d_idx] = 0.0
        
        d_Fdrau[d_idx] = 0.0
        d_Fdrav[d_idx] = 0.0
        d_Fdraw[d_idx] = 0.0
        
        d_Fintu[d_idx] = 0.0
        d_Fintv[d_idx] = 0.0
        d_Fintw[d_idx] = 0.0

    def loop(self, d_idx, d_u, d_v, d_w, d_Vol, d_na, d_dia,
                   d_u_javerage, d_v_javerage, d_w_javerage,
                   d_Fdrau, d_Fdrav, d_Fdraw, d_betaIMEC, 
                   d_Fintu, d_Fintv, d_Fintw, d_Re, d_Cdrag):

        tempu = (d_u_javerage[d_idx] - d_u[d_idx])
        tempv = (d_v_javerage[d_idx] - d_v[d_idx])
        tempw = (d_w_javerage[d_idx] - d_w[d_idx])
        temp = sqrt(tempu*tempu + tempv*tempv + tempw*tempw)
        na = d_na[d_idx]
        dia = d_dia[d_idx]
        Volume = d_Vol[d_idx]
               
        Cdrag = d_Cdrag[d_idx]
        
        if d_na[d_idx] <= 0.8:
            d_betaIMEC[d_idx] = 150.0 * (1-na)**2 / na * self.nu / dia**2 +\
                                1.75*(1-na) * self.rho / dia * temp
        else:
            d_Re[d_idx] = self.rho * na * dia * temp / self.nu
            if d_Re[d_idx] >= 1000.0:
                d_Cdrag[d_idx] = 0.43
            else:
                #d_Re[d_idx] = 1.0 if d_Re[d_idx] < 1.0 else d_Re[d_idx]
                d_Cdrag[d_idx] = 24*(1 + 0.15*d_Re[d_idx]**0.687)/d_Re[d_idx]
            
            d_betaIMEC[d_idx] = 0.75 * Cdrag * na*(1-na)/dia * self.rho *\
                                na**(-2.65) * temp
```

**Comment by prabhuramachandran** [comment]
By default not all properties you setup in a particle array are saved to disk. This is because very often there are many properties that are not useful to save on disk. The particle array's `output_property_arrays` attribute is used to determine which properties are saved.
Use the particle array's `set_output_arrays` method to specify exactly what arrays you wish to save to disk: 
https://pysph.readthedocs.io/en/latest/reference/particle_array.html?#pysph.base.particle_array.ParticleArray.set_output_arrays

Alternatively, if you want to save all properties you can pass `--detailed-output` on the command line to save all properties. Of course, this will increase the size of your output files.

**Reply by hrbyrb** [reply]
Thank you for your reply. 
I have found out the reason for the error, there is a line of code in the equation with a wrong position, resulting in the variable equal to 0.

**Reply by prabhuramachandran** [reply]
Sorry, I realize I missed that you had indeed set the output properties on the particle array.  Could you clarify which line of code in your screenshot was causing the problem?

**Comment by hrbyrb** [comment, marked as answer]
The value of Cdrag in line 471 is passed on line 458, which should actually be placed on line 470. I have to say, it was a very embarrassing mistake, haha

![image](https://user-images.githubusercontent.com/73709907/170476499-0650630f-fa06-4308-857a-f12d1cad64e2.png)

**Reply by prabhuramachandran** [reply]
These sort of mistakes happen to the best of us. :)
